### PR TITLE
Applications: Add new KeyboardMapper application

### PR DESCRIPTION
**Pull request by asliturk**
New app for editing and creating new keymap files.

a

### CODE REVIEW

File: AK/JsonObject.h [hunk 1]
```diff
@@ -136,9 +137,36 @@ template<typename Builder>
...
-        break;
+    case Type::String: {
+        auto size = m_value.as_string->length();
+        builder.append("\"");
+        for (size_t i = 0; i < size; i++) {
+            char ch = m_value.as_string->characters()[i];
+            switch (ch) {
+            case '\e':
+                builder.append("\\e");
+                break;
```
Comment (bugaevc) [inline AK/JsonObject.h:148]: I don't think `\e` is a thing in JSON. You have to use `\u0033` or whatever the proper code.

File: Userland/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,7 @@ target_link_libraries(disasm LibX86)
 target_link_libraries(functrace LibDebug LibX86)
 target_link_libraries(html LibWeb)
 target_link_libraries(ht LibWeb)
+target_link_libraries(keymap LibKeyboard LibCore)
```
Comment (bugaevc) [inline Userland/CMakeLists.txt:17]: No need to add `LibCore` here explicitly, all userland executables already link to it.

File: Libraries/CMakeLists.txt [hunk 3]
```diff
@@ -21,4 +22,4 @@ add_subdirectory(LibThread)
 add_subdirectory(LibTLS)
 add_subdirectory(LibVT)
 add_subdirectory(LibWeb)
-add_subdirectory(LibX86)
+add_subdirectory(LibX86)
```
Comment (bugaevc) [inline Libraries/CMakeLists.txt:25]: ?

File: Libraries/LibKeyboard/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+set(SOURCES
+    CharacterMap.cpp
+    CharacterMapFile.cpp
+)
+
+serenity_libc(LibKeyboard keyboard)
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CMakeLists.txt:6]: How is it a LibC?

File: Libraries/LibKeyboard/CharacterMap.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "CharacterMap.h"
+#include <Kernel/Syscall.h>
+#include <LibKeyboard/CharacterMapFile.h>
+
+namespace LibKeyboard {
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CharacterMap.cpp:31]: The namespace should be `Keyboard`

File: Libraries/LibKeyboard/CharacterMap.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+CharacterMap::CharacterMap(String file_name)
+{
+    auto result = CharacterMapFile::load_from_file(file_name);
+    if (!result.has_value()) {
+        ASSERT_NOT_REACHED();
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CharacterMap.cpp:37]: `ASSERT(result.has_value());`

File: Libraries/LibKeyboard/CharacterMap.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <AK/String.h>
+#include <LibKeyboard/CharacterMapData.h>
+
+namespace LibKeyboard {
+
+class CharacterMap {
+
+public:
+    CharacterMap(String file_name);
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CharacterMap.h:37]:
```suggestion
    CharacterMap(const String& file_name);
```

File: Libraries/LibKeyboard/CharacterMapFile.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    return character_map;
+}
+
+char* CharacterMapFile::read_map(const JsonObject& json, const String& name)
+{
+    if (!json.has(name))
+        return nullptr;
+
+    char* map = new char[CHAR_MAP_SIZE]();
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CharacterMapFile.cpp:79]: You leak this allocation

File: Libraries/LibKeyboard/CharacterMapFile.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    auto map_arr = json.get(name).as_array();
+
+    for (int i = 0; i < map_arr.size(); i++) {
+        auto key_value = map_arr.at(i).as_string();
+        char character = 0;
+        if (key_value.length() == 0) {
+            ;
```
Comment (bugaevc) [inline Libraries/LibKeyboard/CharacterMapFile.cpp:86]: :|

File: Kernel/Devices/KeyboardDevice.cpp [hunk 10]
```diff
@@ -475,6 +475,7 @@ void KeyboardDevice::key_state_changed(u8 raw, bool pressed)
 
     Event event;
     event.key = key;
+    event.scancode =(m_has_e0_prefix) ? 0xe000 + raw : raw;
```
Comment (bugaevc) [inline Kernel/Devices/KeyboardDevice.cpp:478]:
```suggestion
    event.scancode = m_has_e0_prefix ? 0xe000 + raw : raw;
```

File: Kernel/KeyCode.h [hunk 11]
```diff
@@ -160,6 +160,7 @@ enum KeyModifier {
 
 struct KeyEvent {
     KeyCode key { Key_Invalid };
+    u32 scancode {0} ;
```
Comment (bugaevc) [inline Kernel/KeyCode.h:163]:
```suggestion
    u32 scancode { 0 };
```

File: Applications/CMakeLists.txt [hunk 12]
```diff
@@ -9,8 +9,9 @@ add_subdirectory(FontEditor)
 add_subdirectory(Help)
 add_subdirectory(HexEditor)
 add_subdirectory(IRCClient)
-add_subdirectory(Piano)
+add_subdirectory(KeyboardMapper)
 add_subdirectory(PixelPaint)
+add_subdirectory(Piano)
```
Comment (bugaevc) [inline Applications/CMakeLists.txt:14]: Pretty sure Piano sorts before PixelPaint

File: Applications/KeyboardMapper/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,8 @@
+set(SOURCES
+    KeyboardMapperWidget.cpp
+	KeyButton.cpp
+    main.cpp
```
Comment (bugaevc) [inline Applications/KeyboardMapper/CMakeLists.txt:4]: Something happened to alignment here

File: Applications/KeyboardMapper/KeyButton.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    void set_pressed(bool value) { m_pressed = value; }
+
+    Function<void()> on_click;
+
+protected:
+    virtual void click(unsigned modifiers = 0) override;
+    virtual void paint_event(GUI::PaintEvent&) override;
+
+private:
+    explicit KeyButton();
```
Comment (bugaevc) [inline Applications/KeyboardMapper/KeyButton.h:46]: Does this need a constructor?

File: Applications/KeyboardMapper/main.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    // Actions
+    auto open_action = GUI::CommonActions::make_open_action(
+        [&](auto&) {
+            Optional<String> path = GUI::FilePicker::get_open_filepath("Open");
+            if (path.has_value()) {
+                keyboard_mapper_widget->load_from_file(path.value());
+            }
+        });
```
Comment (bugaevc) [inline Applications/KeyboardMapper/main.cpp:71]: Alignment

**Comment by bugaevc** [comment]
I don't like the name LibKeyboard for this. It should be more like LibKeyMap or something. And it doesn't seem to even do much...

File: Libraries/CMakeLists.txt (hunk 3 above)
Comment (asliturk) [reply Libraries/CMakeLists.txt:25]: Last new line added back :(

File: AK/JsonObject.h (hunk 1 above)
Comment (asliturk) [reply AK/JsonObject.h:148]: In current files it is used as "\\x1B". I am going to change it with this value.

Comment (bugaevc) [reply AK/JsonObject.h:148]: I don't think JSON supports `\x` escapes either.

File: Applications/KeyboardMapper/main.cpp (hunk 15 above)
Comment (asliturk) [reply Applications/KeyboardMapper/main.cpp:71]: Hi,
When i remove it, i get following error.

```
[0/1] Re-running CMake...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/asliturk/git-repos/serenity/Build
[1/1] Linking CXX executable Userland/keymap
FAILED: Userland/keymap 
: && /home/asliturk/git-repos/serenity/Toolchain/Local/bin/i686-pc-serenity-g++  -Wall -Wextra -Werror -std=c++2a -fdiagnostics-color=always -Os -Wno-sized-deallocation -fno-sized-deallocation -fno-exceptions -fno-rtti -Wno-address-of-packed-member -Wundef -Wcast-qual -Wwrite-strings -Wimplicit-fallthrough -Wno-nonnull-compare -Wno-deprecated-copy -Wno-expansion-to-defined -DDEBUG -DSANITIZE_PTRS  --sysroot /home/asliturk/git-repos/serenity/Build/Root Userland/CMakeFiles/keymap.dir/keymap.cpp.o  -o Userland/keymap -L/home/asliturk/git-repos/serenity/Build/Libraries/LibC   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibM   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread -Wl,-rpath,/home/asliturk/git-repos/serenity/Build/Libraries/LibC:/home/asliturk/git-repos/serenity/Build/Libraries/LibM:/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread:  Libraries/LibCore/libcore.a  Libraries/LibKeyboard/libkeyboard.a  Libraries/LibC/libc.a  Libraries/LibC/libcrt0.a && :
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: Libraries/LibKeyboard/libkeyboard.a(CharacterMapFile.cpp.o): in function `Keyboard::CharacterMapFile::load_from_file(AK::String const&)':
CharacterMapFile.cpp:(.text+0x3a9): undefined reference to `Core::File::File(AK::StringView const&, Core::Object*)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x3d1): undefined reference to `Core::File::open(Core::IODevice::OpenMode)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x41e): undefined reference to `Core::IODevice::error_string() const'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x45e): undefined reference to `Core::IODevice::read_all()'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
 ~
```

File: Userland/CMakeLists.txt (hunk 2 above)
Comment (asliturk) [reply Userland/CMakeLists.txt]: Hi,
When i remove it, i get following error.

```
[0/1] Re-running CMake...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/asliturk/git-repos/serenity/Build
[1/1] Linking CXX executable Userland/keymap
FAILED: Userland/keymap 
: && /home/asliturk/git-repos/serenity/Toolchain/Local/bin/i686-pc-serenity-g++  -Wall -Wextra -Werror -std=c++2a -fdiagnostics-color=always -Os -Wno-sized-deallocation -fno-sized-deallocation -fno-exceptions -fno-rtti -Wno-address-of-packed-member -Wundef -Wcast-qual -Wwrite-strings -Wimplicit-fallthrough -Wno-nonnull-compare -Wno-deprecated-copy -Wno-expansion-to-defined -DDEBUG -DSANITIZE_PTRS  --sysroot /home/asliturk/git-repos/serenity/Build/Root Userland/CMakeFiles/keymap.dir/keymap.cpp.o  -o Userland/keymap -L/home/asliturk/git-repos/serenity/Build/Libraries/LibC   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibM   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread -Wl,-rpath,/home/asliturk/git-repos/serenity/Build/Libraries/LibC:/home/asliturk/git-repos/serenity/Build/Libraries/LibM:/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread:  Libraries/LibCore/libcore.a  Libraries/LibKeyboard/libkeyboard.a  Libraries/LibC/libc.a  Libraries/LibC/libcrt0.a && :
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: Libraries/LibKeyboard/libkeyboard.a(CharacterMapFile.cpp.o): in function `Keyboard::CharacterMapFile::load_from_file(AK::String const&)':
CharacterMapFile.cpp:(.text+0x3a9): undefined reference to `Core::File::File(AK::StringView const&, Core::Object*)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x3d1): undefined reference to `Core::File::open(Core::IODevice::OpenMode)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x41e): undefined reference to `Core::IODevice::error_string() const'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x45e): undefined reference to `Core::IODevice::read_all()'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
 ~
```

Comment (bugaevc) [reply Userland/CMakeLists.txt]: This sounds like LibKeyboard uses LibCore, but doesn't link to it properly.

File: Libraries/LibKeyboard/CharacterMapFile.cpp (hunk 8 above)
Comment (asliturk) [reply Libraries/LibKeyboard/CharacterMapFile.cpp:79]: I try to fix this but i'm not sure for my solution. Can you review this for me?

Comment (bugaevc) [reply Libraries/LibKeyboard/CharacterMapFile.cpp:84]: It looks like you don't leak it anymore, yeah. But still, don't do it like that. Use `OwnPtr<>`, or a `ByteBuffer`, or something that can deallocate itself when you drop it.

File: Libraries/LibKeyboard/CharacterMapFile.cpp (hunk 9 above)
Comment (asliturk) [reply Libraries/LibKeyboard/CharacterMapFile.cpp:91]: I am not happy with this. JSON parser needs more interest for fix this.

File: AK/JsonObject.h (hunk 1 above)
Comment (asliturk) [reply AK/JsonObject.h:148]: Yes, json.org says only \u is correct. I amde changes in JSONParser to handle \u. Can you review it for me?

File: Libraries/LibKeyboard/CharacterMapFile.cpp (hunk 8 above)
Comment (asliturk) [reply Libraries/LibKeyboard/CharacterMapFile.cpp]: `ByteBuffer` is a good for me :+1:

**Comment by asliturk** [comment]
> I don't like the name LibKeyboard for this. It should be more like LibKeyMap or something. And it doesn't seem to even do much...

Hi, current version of library just collect comment for keymap and Keyboard mapper apps. But in the feature may be we move character mapping logic from kernel to here, inline coded key layout (`KeyPosition.h`) and some new APIs for key leds and keyboard settings. So `LibKeyMap` or `LibCharMap` are good alternatives but in my perspective it will be more then this.

**Comment by bugaevc** [comment]
Ah, that makes sense 🙂

Perhaps we could move `KeyCode.h` from the kernel right now?

File: Applications/KeyboardMapper/KeyPositions.h [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    [ 5] = {  0x05, 204,  0,  50,  50,  true,  5, "4"},
+    [ 6] = {  0x06, 255,  0,  50,  50,  true,  6, "5"},
+    [ 7] = {  0x07, 306,  0,  50,  50,  true,  7, "6"},
+    [ 8] = {  0x08, 357,  0,  50,  50,  true,  8, "7"},
+    [ 9] = {  0x09, 408,  0,  50,  50,  true,  9, "8"},
+    [10] = {  0x0A, 459,  0,  50,  50,  true, 10, "9"},
+    [11] = {  0x0B, 510,  0,  50,  50,  true, 11, "0"},
+    [12] = {  0x0C, 561,  0,  50,  50,  true, 12, "-"},
+    [13] = {  0x0D, 612,  0,  50,  50,  true, 13, "="},
+    [14] = {  0x0E, 663,  0, 100,  50, false,  0, "bask space"},
```
Comment (JonArcherII) [inline Applications/KeyboardMapper/KeyPositions.h:58]:
```suggestion
    [14] = {  0x0E, 663,  0, 100,  50, false,  0, "back space"},
```

Comment (asliturk) [reply Applications/KeyboardMapper/KeyPositions.h]: Thanks for the review.

**Comment by asliturk** [comment]
> Ah, that makes sense slightly_smiling_face
> 
> Perhaps we could move `KeyCode.h` from the kernel right now?

It's earlier for that. First we have remove static maps from `KeyboardDevice.cpp` and then move `KeyPosition.h` to a config file to remove keycode name from `KeyCode.h`.

At this point we have to this think how we represent key short cuts like `Ctrl+C` for non-QWERTY keyboard layouts.

File: Applications/KeyboardMapper/KeyButton.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <LibGUI/AbstractButton.h>
+
+class KeyButton : public GUI::AbstractButton {
```
Comment (awesomekling) [inline Applications/KeyboardMapper/KeyButton.h:31]: Do we really need a custom class for this, or can we use a regular `GUI::Button` and change its button color via the palette?
Maybe I'm missing something here.

File: Libraries/LibKeyboard/CharacterMap.h [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+#pragma once
+
+#include <AK/String.h>
+#include <LibKeyboard/CharacterMapData.h>
+
+namespace LibKeyboard {
+
+class CharacterMap {
```
[the remark this answers is not in the document]
Comment (asliturk) [reply Libraries/LibKeyboard/CharacterMap.h:34]: I am planing add more methods to this class you can review on next version on following URL.
[CharacterMap.h on new branch](https://github.com/asliturk/serenity/blob/KeyboardSettings.App/Libraries/LibKeyboard/CharacterMap.h)

File: Applications/KeyboardMapper/KeyButton.h (hunk 17 above)
Comment (asliturk) [reply Applications/KeyboardMapper/KeyButton.h:31]: I have two reason for this class first one which i failed is return key :) i try to create button like real return and the second one is i can't set background color to `Color::Cyan` when pressed :(

**Pull request merged**